BZ-1903943: low latency cpu partitioning chapter lacks info (#36682)
openshift/openshift-docs: modules/cnf-cpu-infra-container.adoc

[id="cnf-cpu-infra-container_{context}"]
= Restricting CPUs for infra and application containers

> Housekeeping and workloads may utilize the central processing units (CPUs) in a way that would impact latency sensitive processes running on those CPUs. Partitioning your workloads on separate CPUs prevents the processes from interfering with each other. By default, CRI-O uses all online CPUs to run infra containers on nodes, which can result in context switches and spikes in latency. 

Reviewer: I am not sure this is accurate now. What the admin does is really partitioning of the cpus and then assigning the processes (sensitive workload, vs. generic processes, workloads, interrupts and the operating system) to those partitions.

Reviewer: It might be easier to understand from slides 8 and 9 here https://static.sched.com/hosted_files/devconfcz2021/70/Openshift-for-low-latency-and-real-time-workloads.pdf

Author: Thanks! I'll make the suggested changes. I'm also making slide 9 into a table, which might be easier for users to follow.

> The exact partitioning pattern to use depends on many factors like hardware, workload characteristics and system load. Some sample use cases are as follows:

Reviewer: ```suggestion
The exact partitioning pattern to use depends on many factors like hardware, workload characteristics and the expected system load. Some sample use cases are as follows:
```

Author: Updated as requested.

* If the latency sensitive workload uses specific hardware (like network interface cards (NICs)), ensure that the CPUs in the isolated pool are as "close" to this hardware as possible. Place the workload in at least the same Non-Uniform Memory Access (NUMA) node.

> * The reserved pool is used for handling all interrupts, but not all interrupt requests generate noise. When depending on system networking, allocate a sufficiently sized reserve pool to handle all the incoming packet interrupts.

Reviewer: Actually, all interrupts generate noise, but sometimes you just can't avoid it (system networking). Also this would be accurate for 4.6, but 4.7+ is a bit smarter and the user can decide which workloads are sensitive and which are not. See this section: https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus_cnf-master

Author: Updated as requested.

> The above examples require detailed analysis and measurements to be combined into a pattern where all housekeeping CPUs are allocated from a single NUMA node. The reserved CPUs can also be distributed among all NUMA nodes.

Reviewer: Housekeeping can also span multiple NUMA nodes and quite commonly do. There really are both options and it the selection depends on the workload architecture and the specific user use case.

Author: Updated as requested.

[NOTE]

> ====

Reviewer: This is a very important note and one of the typical configuration bugs, can we make it VERY visible?

Author: Yes, absolutely! I can use the "important" or "warning" admonishment instead of "note". I can also move it elsewhere in the document if you'd like.

The reserved and isolated CPU pools must not overlap and together must span all available cores in the worker node.
====

> To ensure that housekeeping tasks and workloads run on separate NUMA nodes, specify two groups of CPUs in the `spec` section of the performance profile. 

Reviewer: ```suggestion
To ensure that housekeeping tasks and workloads do not interfere with each other, specify two groups of CPUs in the `spec` section of the performance profile. 
```

Author: Updated as requested.

* `isolated` - The CPUs for the application container workloads. These CPUs have the lowest latency. Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth.

> * `reserved` - The CPUs for the cluster and operating system housekeeping duties, including pod infra containers. Threads in the `reserved` group tend to be under higher load, so latency-sensitive applications should be run in the `isolated` group. See link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed[Create a pod that gets assigned a QoS class of `Guaranteed`].

Reviewer: The change here is not entirely accurate. The threads are not necessarily overloaded, but there is a lot of "churn" on them. Lots of random processes.

Author: Updated as requested.

[id="cnf-cpu-infra-container_{context}"]
= Restricting CPUs for infra and application containers

> Housekeeping and workloads may use the central processing units (CPUs) in a way that would impact latency sensitive processes running on those CPUs. Partitioning the CPUs prevents the processes from interfering with each other. By default, CRI-O uses all online CPUs to run infra containers on nodes, which can result in context switches and spikes in latency. Use the following table see where process' CPUs run after tuning the node using the Performance Addon Operator:

Reviewer 2: First sentence sounds passive with 'would impact'. I would also hyphenate 'latency sensitive'.
So rewrite as '... in a way that impacts latency-sensitive processes...'
Fourth sentence is missing 'see'.
So rewrite as 'Use the following table to see where...'

Author: Updated as suggested.

> Housekeeping and workloads may use the central processing units (CPUs) in a way that would impact latency sensitive processes running on those CPUs. Partitioning the CPUs prevents the processes from interfering with each other. By default, CRI-O uses all online CPUs to run infra containers on nodes, which can result in context switches and spikes in latency. Use the following table see where process' CPUs run after tuning the node using the Performance Addon Operator:

Author: Updated as suggested.

.Process' CPU assignments
[%header,cols=2*] 
|===
|Process type
|Details

|Burstable and best-effort pods

> |Run on any CPU not used by a low latency container, including reserved

Reviewer: `cpu.reserved` field of the Performance Profile. Which means almost the same thing as system-reserved in the link you posted.

> |Run on any CPU not used by a low latency container, including reserved

Reviewer 3: Ok cool. 

What about: "Runs on any CPU not used by a low latency container, including CPUs as specified in the `cpu.reserved` field of the configured Performance Profile CR"

> |Run on any CPU not used by a low latency container, including reserved

Reviewer: I would personally low-key the including... part. It causes confusion when people think about it. It is technically correct, but harder to explain.

> |Run on any CPU not used by a low latency container, including reserved

Reviewer 3: "Runs on any CPU not used by a low-latency container, such as CPUs specified in the `cpu.reserved` field of the configured Performance Profile CR"

?

> |Run on any CPU not used by a low latency container, including reserved

Author: @MarSik what do you think about Aidan's suggested rewrite here? It makes sense to me, but I'd like your thoughts before I make the changes. Thank you!

> |Run on any CPU not used by a low latency container, including reserved

Reviewer 3: Bowing to @MarSik here, I'm no expert...

> |Run on any CPU not used by a low latency container, including reserved

Author: Received reply from Martin via Slack:

9/28/21 10:02am
> [Aidan's proposed rewrite for line 17 of cnf-cpu-infra-container.adoc] gives exactly opposite impression than what I want. It mentions reserved as a notable group of cpus. What about keeping only the first part and removing all mentions of reserved? Paraphrasing: Run everywhere except where low latency workload is running.

> |Run on any CPU not used by a low latency container, including reserved

Author: Updated as suggested.

> |Runs on any CPU not used by a low latency container, including reserved

Reviewer: Shouldn't it be just Run since the process type content is in plural? (and the same for the rest of the rows)

Author: That it should! Thanks for catching that!

Author: Updated as suggested.

|Infrastructure pods
|Run on any CPU not used by a low latency container, including reserved

|Interrupts

> |Redirected to reserved CPUs (optional in {product-version} and later)

Reviewer 2: The Red Hat Supplementary Style guide uses 'versions' with earlier/later
So rewrite as '...{product-version} and later versions)'

Author: Updated as suggested.

> |Redirected to reserved CPUs (optional in {product-version} and later)

Reviewer 3: I think you should use `{product-title} {product-version}` here. Also suggeted rewrite: "Redirects to reserved CPUs (optional in {product-title} {product-version} and later)"

> |Redirected to reserved CPUs (optional in {product-version} and later)

Author: Updated as suggested.

|Kernel processes
|Pinned to reserved CPUs

|Latency sensitive workload pods
|Pinned to a specific set of exclusive CPUs from the isolated pool

|OS processes/systemd services
|Pinned to reserved CPUs
|===

The exact partitioning pattern to use depends on many factors like hardware, workload characteristics and the expected system load. Some sample use cases are as follows:

> * If the latency sensitive workload uses specific hardware (like network interface cards (NICs)), ensure that the CPUs in the isolated pool are as "close" to this hardware as possible. Place the workload in at least the same Non-Uniform Memory Access (NUMA) node.

Reviewer 2: Like above, I would hyphenate 'latency sensitive'.
So rewrite as 'If the latency-sensitive workload...'

Author: Updated as suggested.

> * If the latency sensitive workload uses specific hardware (like network interface cards (NICs)), ensure that the CPUs in the isolated pool are as "close" to this hardware as possible. Place the workload in at least the same Non-Uniform Memory Access (NUMA) node.

Reviewer 3: Suggest rewriting this sentence to lose the double parenthesis for clarity. Also I don't think you need to quote "close". 

"If the latency sensitive workload uses specific hardware, such as a network interface card (NIC), ensure that the CPUs in the isolated pool are as close as possible to this hardware. At a minimum, you should place the workload in the same Non-Uniform Memory Access (NUMA) node."

Author: Updated as suggested.

> |Latency sensitive workload pods

Reviewer 4: Other renditions of this are latency-sensitive

Author: Updated as suggested.

> * The reserved pool is used for handling all interrupts. When depending on system networking, allocate a sufficiently sized reserve pool to handle all the incoming packet interrupts. In {product-version} and later, workloads can optionally be labeled as sensitive.

Reviewer 2: Second sentence I would hyphenate 'sufficiently sized'.
So rewrite as '...allocate a sufficiently-sized reserve pool...'
The Red Hat Supplementary Style guide uses 'versions' with earlier/later
So rewrite as 'In {product-version} and later versions...'

Author: Updated as suggested.

> The decision regarding which specific CPUs should be used for reserved and isolated partitions requires a detailed analysis and measurements. Factors like NUMA affinity of devices and memory play a role. The selection also depends on the workload architecture and the specific use case.

Reviewer 2: First sentence does read right with ''a' detailed analysis and measurements.'
Do you mean '...reserved and isolated partitions requires detailed analysis and measurements.'
Also, do you need to specify what 'units' are being measured?

Reviewer: In other (engineering:) words it needs to say: Assigning cpus into one of the reserved or isolated pools requires detailed analysis because of workload and hw interactions only the user knows.

The measurement unit is kind of implied by the title of this whole document (low latency = time) so I do not think it is necessary to be that explicit here.

Reviewer 3: What about:
"You decide whether to assign CPUs into reserved or isolated pools by doing a detailed analysis of workload and hardware interactions."

Author: Updated as suggested by Padraig. Did not include Aidan's "You decide" verbiage, because I thought we weren't supposed to use "you" statements.

> * The reserved pool is used for handling all interrupts. When depending on system networking, allocate a sufficiently sized reserve pool to handle all the incoming packet interrupts. In {product-version} and later, workloads can optionally be labeled as sensitive. For more information, see the Red Hat Knowledgebase solution xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus[Managing device interrupt processing for guaranteed pod isolated CPUs].

Reviewer: ```suggestion
* The reserved pool is used for handling all interrupts. When depending on system networking, allocate a sufficiently sized reserve pool to handle all the incoming packet interrupts. In {product-version} and later, workloads can optionally be labeled as sensitive. For more information, see the xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus[Managing device interrupt processing for guaranteed pod isolated CPUs] section of this document.
```

Reviewer: It renders to an internal link within the same doc, also this is the documentation, not knowledge base.

Author: Good catch! Updated the link to reflect the correct location.

[IMPORTANT]
====
The reserved and isolated CPU pools must not overlap and together must span all available cores in the worker node.
====

To ensure that housekeeping tasks and workloads do not interfere with each other, specify two groups of CPUs in the `spec` section of the performance profile. 

> * `isolated` - The CPUs for the application container workloads. These CPUs have the lowest latency. Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth.

Reviewer 3: ```suggestion
* `isolated` - Specifies the CPUs for the application container workloads. These CPUs have the lowest latency. Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth.
```

Author: Update as suggested.

> * `reserved` - The CPUs for the cluster and operating system housekeeping duties. Threads in the `reserved` group are often busy, so latency-sensitive applications should be run in the `isolated` group.

Reviewer 3: ```suggestion
* `reserved` - Specifies the CPUs for the cluster and operating system housekeeping duties. Threads in the `reserved` group are often busy. Do not run latency-sensitive applications should be run in the `reserved` group.
```

Reviewer: @aireilly The end of the suggested sentence does not make sense.

Reviewer 3: Oops! Monday morning low caffeine.

```
* `reserved` - Specifies the CPUs for the cluster and operating system housekeeping duties. Threads in the `reserved` group are often busy. Do not run latency-sensitive applications in the `reserved` group. Latency-sensitive applications should be run in the `isolated` group.
```

Reviewer: I would actually remove `should be` as that implies the user can control and fix it :) But it is the platform doing this for him.

> * `reserved` - The CPUs for the cluster and operating system housekeeping duties, including pod infra containers. Threads in the `reserved` group are often busy, so latency-sensitive applications should be run in the `isolated` group.

Reviewer: ```suggestion
* `reserved` - The CPUs for the cluster and operating system housekeeping duties. Threads in the `reserved` group are often busy, so latency-sensitive applications should be run in the `isolated` group.
```

The infra containers are actually running and accounting towards the isolated pool. That they can physically execute on the reserved cores is a bit confusing and just an internal mechanics that does not change much.

Author: Updated as suggested.

[id="cnf-cpu-infra-container_{context}"]
= Restricting CPUs for infra and application containers

> Housekeeping and workloads may use the central processing units (CPUs) in a way that would impact latency sensitive processes running on those CPUs. Partitioning the CPUs and assigning the processes prevents the processes from interfering with each other. By default, CRI-O uses all online CPUs to run infra containers on nodes, which can result in context switches and spikes in latency. Use the following table to assign processes:

Reviewer: ```suggestion
Housekeeping and workloads may use the central processing units (CPUs) in a way that would impact latency sensitive processes running on those CPUs. Partitioning the CPUs and assigning the processes prevents the processes from interfering with each other. By default, CRI-O uses all online CPUs to run infra containers on nodes, which can result in context switches and spikes in latency. Check out the following table to see where processes run after tuning the node using the Performance Addon Operator:
```

Reviewer: The user does not pin the processes himself, he just decides which CPUs are reserved and isolated and the platform does the pinning for him.

Author: Updated as suggested.

.Process assignments
[%header,cols=3*] 
|===

> |Core	

Reviewer: The Core column does not make sense. Please remove it.

Author: Removed Core column as requested.

> |Process

Reviewer: ```suggestion
|Process type
```
Just a suggestion, there are many different interrupts, kernel processes etc.

Author: Changing to "Process type" for clarity.

|Details

|0
|Interrupts
|Redirected to reserved CPUs (optional in 4.7+)

|1
|Kernel processes
|Pinned to reserved CPUs

|2
|OS processes/systemd services
|Pinned to reserved CPUs

|3
|Infrastructure pods
|Runs on any CPU not used by a low latency container, including reserved

|4

> |General workflow pod

Reviewer: ```suggestion
|Burstable and best-effort pods
```

Author: Updated as suggested.

|Runs on any CPU not used by a low latency container, including reserved

|5
|Latency sensitive workload pods
|Pinned to a specific set of exclusive CPUs from the isolated pool
|===

The exact partitioning pattern to use depends on many factors like hardware, workload characteristics and the expected system load. Some sample use cases are as follows:

* If the latency sensitive workload uses specific hardware (like network interface cards (NICs)), ensure that the CPUs in the isolated pool are as "close" to this hardware as possible. Place the workload in at least the same Non-Uniform Memory Access (NUMA) node.

> * The reserved pool is used for handling all interrupts. When depending on system networking, allocate a sufficiently sized reserve pool to handle all the incoming packet interrupts. In version 4.7+, workloads can optionally be labeled as sensitive. More information regarding the management of device interrupt processing can be found in link:https://docs.openshift.com/container-platform/4.8/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.html#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus_cnf-master[Managing device interrupt processing for guaranteed pod isolated cpus].

Reviewer: You should probably make the link relative only or point to the same version of docs using some macro like {version} (?)

Author: Ah yes, a macro would make the most sense! Thanks for pointing that out.

Author: Updated to "In {product-version} and later" and updated link to reflect xref and .adoc:

> xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus[Managing device interrupt processing for guaranteed pod isolated CPUs]

> The above examples require detailed analysis and measurements to be combined into a pattern where all housekeeping CPUs are allocated from the NUMA node(s). The reserved CPUs can also be distributed among all NUMA nodes. The selection depends on the workload architecture and the specific use case.

Reviewer: ```suggestion
The decision about which specific CPUs should be used for reserved and isolated partitions requires a detailed analysis and measurements. Factors like NUMA affinity of devices and memory play a role. The selection also depends on the workload architecture and the specific use case.
```

Author: Updated as suggested.

[IMPORTANT]
====
The reserved and isolated CPU pools must not overlap and together must span all available cores in the worker node.
====

To ensure that housekeeping tasks and workloads do not interfere with each other, specify two groups of CPUs in the `spec` section of the performance profile. 

* `isolated` - The CPUs for the application container workloads. These CPUs have the lowest latency. Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth.

> * `reserved` - The CPUs for the cluster and operating system housekeeping duties, including pod infra containers. Threads in the `reserved` group tend to run multiple processes at random, so latency-sensitive applications should be run in the `isolated` group. See link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed[Create a pod that gets assigned a QoS class of `Guaranteed`].

Reviewer: I still like the old "busy" work better :)

Author: Updated as suggested.

<2> Specify which CPUs are for application containers to run workloads. 
<3> Optional: Specify a node selector to apply the performance profile to specific nodes.

.Additional information

> link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed[Create a pod that gets assigned a QoS class of Guaranteed]

Reviewer 5: The heading for this should be "Additional resources" (not "information"), and that anything under it should always be a bullet, even if it's just one. But it looks like you already have one under it in the assembly, so it's appearing back-to-back in the rendering. Can you move this link to there?

Author: Updated as suggested.
